### PR TITLE
Add Rose optimizer

### PR DESCRIPTION
**Pull request by MatthewK78**
Rose optimizer: https://github.com/MatthewK78/Rose

- Stateless: uses less memory than even 8-bit versions of other optimizers. If not for temporary working memory, its memory usage would be as low as plain vanilla SGD (**without** momentum).
- Matches or exceeds AdamW in observed accuracy and convergence.
- Training loss runs higher while validation loss runs lower, consistent with implicit regularization rather than memorization.

Note: Requires a higher than usual learning rate. For example, 1e-3 instead of AdamW's 1e-4.

### CODE REVIEW

**Comment by neph1** [comment]
I can't vouch for the results, but I did my own implementation of Rose and it worked, at least :)
I'll do a training run with this PR and see how it comes out (can't promise comparisons).

**Comment by neph1** [comment]
I did a quick flux 2 klein test, 200 iterations, lr = 3e-3, and it can say that it works. I tested at 100 iterations, but it was not finished. Whether "better" or not is difficult to say, because I didn't do any side by sides, and it's so subjective anyway.

I also used: 
`compute_dtype = 'fp32'`
because that was recommended in the original post. Maybe not relevant here.

**Comment by MatthewK78** [comment]
> I did a quick flux 2 klein test, 200 iterations, lr = 3e-3, and it can say that it works. I tested at 100 iterations, but it was not finished. Whether "better" or not is difficult to say, because I didn't do any side by sides, and it's so subjective anyway.
> 
> I also used: `compute_dtype = 'fp32'` because that was recommended in the original post. Maybe not relevant here.

Thanks for trying it out! 😊 It trains a bit differently from momentum-based optimizers and can take some experimentation to get used to it. In some cases, convergence can happen significantly faster than Adam. Overfitting tends to look different, so if you experience strange artifacting or other issues you may want to try an earlier checkpoint. Another thing to watch out for is expecting training loss to mirror Adam; it can be quite a bit higher even though validation loss is lower.

3e-3 might be too high without an LR scheduler. If you notice wonky distortions you might want to try bumping that down a bit. Also, I noticed the current cosine LR scheduler implementation doesn't have a way to specify `eta_min` (it's hard-coded at 1e-6), which means training progress might stop too soon.

Hmm, I'm not sure which post you're referring to by "original post." fp64 is the default and still results in using significantly less VRAM than AdamW8bit. With stateful optimizers fp64 would be impractical, but since Rose is stateless, using fp64 precision has a small impact on memory/performance and gives a nice boost to accuracy. I would only use a `compute_dtype` of fp32/bf16 if absolutely necessary.

**Comment by neph1** [comment]
I've been experimenting some more with it. I wanted to try a style lora and it took a bit longer than I had hoped. Here is the result: [https://civitai.com/models/1371819/1920s-horror-movies-lora](https://civitai.com/models/1371819/1920s-horror-movies-lora?modelVersionId=3049887)
It's at 80 epochs / 560 steps. It's marginally better than the checkpoint at 60 epochs, but a bit better in style.

Here are the optimizer settings:
```
[optimizer]
type = 'Rose'
lr = 3e-3
weight_decay = 0.01
centralize = true
stabilize = true
bf16_sr = true
compute_dtype = 'fp64'
wd_schedule = false
```
I always use a linear lr scheduler. It's obvious to me, but not to others :)
Maybe a bit higher falloff or lower starting lr could have improved it further (at least when looking at the loss).
I didn't get a chance to look at the VRAM use compared to my older training runs.

Regarding any other assumptions I had made, I probably remembered things wrong.

It's nice to see something new come along :)

**Comment by MatthewK78** [comment]
@neph1 Very cool, thanks for sharing! I like the lora, looks good! You may want to consider enabling `wd_schedule` if you're using any lr scheduler.

**Comment by LorisTecnology** [comment]
@neph1 how have you been able to implement it into Diffusion Pipe? i want to test as well

**Comment by neph1** [comment]
It seems I copied the whole class into optimizers and then similarly to the PR:

```
        elif optim_type_lower == 'rose':
            from optimizers import rose
            klass = rose.Rose
```

Simplest is probably installing it with pip and do the same as in this pr.

**Comment by MatthewK78** [comment]
Yep, as @neph1 said, doing a pip install is easier, especially considering I'm still working on improving the optimizer and a simple pip upgrade in the future will be painless. In fact, I recently managed to achieve even better results with some promising new enhancements I'm currently testing.

**Comment by LorisTecnology** [comment]
tested and works really well. On Krea 2 at epoch15
